stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,233 @@
+#ifndef STAN_MATH_REV_FUN_HMM_MARGINAL_LPDF_HPP
+#define STAN_MATH_REV_FUN_HMM_MARGINAL_LPDF_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/row.hpp>
+#include <stan/math/prim/fun/col.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/prim/fun/exp.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+#include <stan/math/prim/core.hpp>
+#include <Eigen/Core>
+#include <vector>
+#include <iostream>
+
+namespace stan {
+namespace math {
+
+/**
+ * For a Hidden Markov Model with observation y, hidden state x,
+ * and parameters theta, return the log marginal density, log
+ * p(y | theta). In this setting, the hidden states are discrete
+ * and take values over the finite space {1, ..., K}.
+ * The marginal lpdf is obtained via a forward pass.
+ * The [in, out] argument are saved so that we can use them when
+ * calculating the derivatives.
+ *
+ * @param[in] log_omega log matrix of observational densities.
+ *              The (i, j)th entry corresponds to the
+ *              density of the ith observation, y_i,
+ *              given x_i = j.
+ * @param[in] Gamma transition density between hidden states.
+ *              The (i, j)th entry is the probability that x_n = j,
+ *              given x_{n - 1} = i. The rows of Gamma are simplexes.
+ * @param[in] rho initial state
+ * @param[in, out] alphas unnormalized partial marginal density.
+ *                   The jth column is the joint density over all
+ *                   observations y and the hidden state j.
+ * @param[in, out] alpha_log_norms max coefficient for column of alpha,
+ *                   to be used to normalize alphas.
+ * @param[in, out] omegas term-wise exponential of omegas.
+ * @return log marginal density.
+ */
+double hmm_marginal_lpdf(const Eigen::MatrixXd& log_omegas,
+                         const Eigen::MatrixXd& Gamma,
+                         const Eigen::VectorXd& rho, Eigen::MatrixXd& alphas,
+                         Eigen::VectorXd& alpha_log_norms,
+                         Eigen::MatrixXd& omegas) {
+  omegas = log_omegas.array().exp();  // CHECK -- why the .array()?
+  int n_states = log_omegas.rows();
+  int n_transitions = log_omegas.cols() - 1;
```

Comment:
hmmm... overkilled, no?

Reply:
No it's good to let the compiler know things are const

Reply:
It also let's readers know "The value of this object will not change through the rest of it's life"

Reply:
hmmm.... I'm going to need a high charisma roll to be convinced.

Reply:
My dude you can chase your bliss and if that doesn't involve const then sure that's fine. But if you don't mutate something later why not let the reader know it won't change by declaring it const?

Reply:
Mainly because it clutters the code.  The compiler can't act just on `const` because it's possible to cast away the `const` modifier.  I try to follow the lead of mature packages like Boost, which don't declare all their local variables (or arithmetic function args) `const`.

Reply:
Ok, that's settled then.